motionplan/constraint_handler.go

```diff
@@ -0,0 +1,575 @@
+package motionplan
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/pkg/errors"
+
+	"go.viam.com/rdk/pointcloud"
+	"go.viam.com/rdk/referenceframe"
+	"go.viam.com/rdk/spatialmath"
+)
+
+var defaultMinStepCount = 2
+
+// ConstraintHandler is a convenient wrapper for constraint handling which is likely to be common among most motion
+// planners. Including a constraint handler as an anonymous struct member allows reuse.
+type ConstraintHandler struct {
+	segmentConstraints   map[string]SegmentConstraint
+	segmentFSConstraints map[string]SegmentFSConstraint
+	stateConstraints     map[string]StateConstraint
+	stateFSConstraints   map[string]StateFSConstraint
+	pathMetric           StateFSMetric // Distance function which converges on the valid manifold of intermediate path states
+	boundingRegions      []spatialmath.Geometry
+}
+
+// NewEmptyConstraintHandler - creates a ConstraintHandler with nothing.
+func NewEmptyConstraintHandler() *ConstraintHandler {
+	handler := ConstraintHandler{}
+	handler.pathMetric = NewZeroFSMetric()
+	return &handler
+}
+
+// NewConstraintHandlerWithPathMetric - creates a ConstraintHandler with a specific metric.
+func NewConstraintHandlerWithPathMetric(m StateFSMetric) *ConstraintHandler {
+	handler := ConstraintHandler{}
+	handler.pathMetric = m
+	return &handler
+}
+
+// NewConstraintHandler - creates a ConstraintHandler with all the params.
+func NewConstraintHandler(
+	collisionBufferMM float64,
+	constraints *Constraints,
+	startPoses, goalPoses referenceframe.FrameSystemPoses,
+	fs *referenceframe.FrameSystem,
+	movingRobotGeometries, staticRobotGeometries []spatialmath.Geometry,
+	seedMap referenceframe.FrameSystemInputs,
+	worldState *referenceframe.WorldState,
+	boundingRegions []spatialmath.Geometry,
+	useTPspace bool,
+) (*ConstraintHandler, error) {
+	if constraints == nil {
+		// Constraints may be nil, but if a motion profile is set in planningOpts
+		// we need it to be a valid pointer to an empty struct.
+		constraints = &Constraints{}
+	}
+	handler := NewEmptyConstraintHandler()
+	handler.boundingRegions = boundingRegions
+
+	frameSystemGeometries, err := referenceframe.FrameSystemGeometries(fs, seedMap)
+	if err != nil {
+		return nil, err
+	}
+
+	obstaclesInFrame, err := worldState.ObstaclesInWorldFrame(fs, seedMap)
+	if err != nil {
+		return nil, err
+	}
+	worldGeometries := obstaclesInFrame.Geometries()
+
+	frameNames := map[string]bool{}
+	for _, fName := range fs.FrameNames() {
+		frameNames[fName] = true
+	}
+
+	allowedCollisions, err := collisionSpecifications(
+		constraints.GetCollisionSpecification(),
+		frameSystemGeometries,
+		frameNames,
+		worldState.ObstacleNames(),
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	if useTPspace {
```

Comment:
I saw this got added back in in its own commit. Is this a good time to ask what tp (topological?) space is and when we use it?

Or is this just more of a "test called a method with tp space = true and we did the wrong thing" and I shouldn't read any more into this at this time?

Reply:
More the latter.
It's only used for base planning and slam, so not really needed now, just trying to rationalize the world